### PR TITLE
Upstream simplified Optimism test

### PR DESCRIPTION
**Pull request by palinatolmach**
Closes https://github.com/runtimeverification/kontrol/issues/366.

This PR upstreams a simplified Optimism test [PortalTest](https://github.com/runtimeverification/kontrol/blob/aace1d2eac9edb2b4a031e78ee59103499d66b18/src/tests/integration/test-data/foundry/test/PortalTest.sol) that can be used for performance profiling:
```solidity
/// @custom:kontrol-array-length-equals _withdrawalProof: 1,
/// @custom:kontrol-bytes-length-equals _withdrawalProof: 32,
function test_withdrawal_paused(
    Types.WithdrawalTransaction memory _tx,
    uint256 _l2OutputIndex,
    Types.OutputRootProof calldata _outputRootProof,
    bytes[] calldata _withdrawalProof
)
    external
{
    vm.expectRevert();
    portalContract.proveWithdrawalTransaction(_tx, _l2OutputIndex, _outputRootProof, _withdrawalProof);
}
```
It also adds a set of lemmas needed for this test ([pausability-lemmas.k](https://github.com/runtimeverification/kontrol/blob/upstream-op-test/src/tests/integration/test-data/pausability-lemmas.k)), and renames an existing file with lemmas for the `sumToN` test ([sum-to-n-lemmas.k](https://github.com/runtimeverification/kontrol/blob/b854b57716179670eb94633eb8a4614210e0926d/src/tests/integration/test-data/sum-to-n-lemmas.k)).

The execution of this test on my machine takes: 
- 5m19s using latest (v0.1.204) Kontrol
-  7m40s using v0.1.178
-  15m15s using v0.1.176
-  11m40s using v0.1.176 with booster overwritten to [d299753](https://github.com/runtimeverification/hs-backend-booster/pull/503) (with simplifier cache enabled)

### CODE REVIEW

**Comment by palinatolmach** [comment]
The CI job with a new test got killed (https://github.com/runtimeverification/kontrol/actions/runs/7898491231/attempts/1?pr=369, re-running it now), and on my machine it also took ~an hour — I'll look into how it can be further minimized while still being illustrative of the performance changes.

**Comment by palinatolmach** [comment]
The following test (which is too simple to be illustrative of the performance difference between backend versions):
```solidity
    /// @custom:kontrol-length-equals _withdrawalProof: 3,
    /// @custom:kontrol-length-equals data: 32,
    /// @custom:kontrol-length-equals _withdrawalProof[]: 32,
    function test_withdrawal_paused(
        Types.WithdrawalTransaction calldata _tx,
        uint256 _l2OutputIndex,
        Types.OutputRootProof calldata _outputRootProof,
        bytes[] calldata _withdrawalProof
    )
        external
    {
        portalContract.pause();
        vm.expectRevert();
        portalContract.proveWithdrawalTransaction(_tx, _l2OutputIndex, _outputRootProof, _withdrawalProof);
    }
}
```
takes ~6 minutes to run locally on my machine and ~8 minutes — on CI when executed by a faster runner. It seems that we're getting very different execution time on booster tests depending on the runner, e.g.,
- github-runner-2 consistently took more than 2 hours to run a job: https://github.com/runtimeverification/kontrol/actions/runs/7914368796/job/21642139595, while
- github-runner-36 finished it in 1h8m: https://github.com/runtimeverification/kontrol/actions/runs/7927923233/job/21645283356

Making the test more challenging (and lemma-dependent) is needed to ensure that we can observe the difference in performance, however, it seems that it would also cause time outs in many CI runs. With @JuanCoRo, we've also realised that a cleanup of lemmas used to run this test is needed, and it seems that removing some of the lemmas improves the execution time w/o causing any issues — that's what I'm looking into now.

**Comment by palinatolmach** [comment]
I've pushed an updated [test](https://github.com/runtimeverification/kontrol/blob/0c4ab95f45d7d290aa69c9c677a10b1fe4617385/src/tests/integration/test-data/foundry/test/OptimismPortalKontrol.t.sol) as well as contracts ([OptimismPortal](https://github.com/runtimeverification/kontrol/blob/0c4ab95f45d7d290aa69c9c677a10b1fe4617385/src/tests/integration/test-data/foundry/src/OptimismPortal.sol), [Types](https://github.com/runtimeverification/kontrol/blob/0c4ab95f45d7d290aa69c9c677a10b1fe4617385/src/tests/integration/test-data/foundry/src/libraries/Types.sol)) it's calling. The test demonstrates ~10% performance diffs between Kontrol/backend versions taking
- 12m using latest (v0.1.178) Kontrol
- 10m50s using v0.1.176 (before the recent performance regression got introduced)
- 9m55s using v0.1.176 with booster overwritten to https://github.com/runtimeverification/hs-backend-booster/pull/503 (with simplifier cache enabled)

However, the test is the current format uses a different `solc` version from the rest of the test suite (0.8.15) and can be minimized further. These changes (like any other changes to the Solidity code) require modification of the `copy-memory-to-memory` lemma, that includes the following steps (instruction courtesy of @PetarMax):
- when hitting the first branching, get to 200 steps or so backward
- execute with `--break-every-step` for ~200 iterations, which should reveal the copying loop
- update the rule according to where the loop starts, where it ends, and what is the word stack structure at both of those points

**Review by anvacaru — APPROVED** [review]

File: src/tests/integration/test-data/foundry/src/Portal.sol [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+contract Portal  {
+    bool paused;
+
+    /// @notice Emitted when a withdrawal transaction is proven.
+    event WithdrawalProven(address indexed from, address indexed to);
+
+    /// @notice Reverts when paused.
+    modifier whenNotPaused() {
+        require(paused == false, "OptimismPortal: paused");
```
[the remark this answers is not in the document]
Comment (palinatolmach) [reply src/tests/integration/test-data/foundry/src/Portal.sol]: Thanks! Fixed in https://github.com/runtimeverification/kontrol/pull/369/commits/eb4aeb96315294304514a635979f01d175815022 (the lemma seems unaffected)

**Comment by ehildenb** [comment]
@palinatolmach not sure, but can we get this merged? One of the next tasks for Qian is to investigate loop invariant inference using the `merge-nodes` functionality, so perhaps having this merged will make it easier for her. https://github.com/runtimeverification/kontrol/issues/448

**Comment by palinatolmach** [comment]
Removed `automerge` to let https://github.com/runtimeverification/kontrol/pull/355 go through first.

I couldn't remove the lemmas causing the rewrite in another contract w/o affecting `PortalTest` performance, but the related investigation is factored out into https://github.com/runtimeverification/kontrol/issues/442. Otherwise, it's ready to be merged.

**Pull request merged**